File: Crawler.php
<?php
/*
 * Copyright (c) 2020 Nozomu Yamazaki
 * Released under the MIT license
 * https://opensource.org/licenses/mit-license.php
 * 
 * Crawler Class
 */

// $url = "https://www.chorkleines.com/member/";

if (php_sapi_name() != 'cli') {
    throw new Exception('This application must be run on the command line.');
}

// $url = "https://www.chorkleines.com/member/";
$url = "https://www.chorkleines.com/member/download/18/presta/";
$crawler = new Crawler($url);

class Crawler
{
    private $memory;
    function __construct($url)
    {
        // memorize the document which are already indexed and also new
        $this->memory = array();
        require(__DIR__ . '/core/dbconnect.php');
        $target = date("Y/m/d H:i:s", strtotime("-30 days", strtotime(date("Y/m/d"))));
        $query = "SELECT * FROM documents WHERE last_index IS NOT NULL AND last_index > '$target'";
        $result = $mysqli->query($query);
        if (!$result) {
            print('<strong>Query failed:</strong> ' . $mysqli->error . 'thrown in <strong>' . __FILE__ . '</strong> on line <strong>' . __LINE__ . '</strong>');
            $mysqli->close();
            exit();
        }
        while ($row = $result->fetch_assoc()) {
            array_push($this->memory, $row['url']);
        }
        $this->crawler($url, "", 0);
    }

    /**
     * crawler
     * @param string $url
     * @param int $depth
     */
    private function crawler(string $url, string $title, int $depth)
    {
        require_once(__DIR__ . "/get_html.php");
        require_once(__DIR__ . "/mecab.php");
        require_once(__DIR__ . "/pdf_to_text.php");
        require_once(__DIR__ . "/html_to_text.php");
        require_once(__DIR__ . "/insert_document.php");
        require_once(__DIR__ . "/insert_word.php");
        require_once(__DIR__ . "/index_finished.php");
        require_once(__DIR__ . "/get_absolute_paths.php");
        require(__DIR__ . '/core/dbconnect.php');

        if (strpos($url, "chorkleines.com/member/") === false) {
            return;
        } elseif (strpos($url, "chorkleines.com/member/bbs/") !== false) {
            return;
        } elseif (strpos($url, "chorkleines.com/member/download/18/pdf_search/") !== false) {
            return;
        } elseif (strpos($url, "chorkleines.com/member/download/18/scoredb/") !== false) {
            return;
        } elseif (strpos($url, "chorkleines.com/member/download/18/past_exam/") !== false) {
            return;
        } elseif (strpos($url, "chorkleines.com/member/wiki/") !== false) {
            return;
        } elseif (strpos($url, "chorkleines.com/member/kleines_search/") !== false) {
            return;
        }

        // when the file is pdf
        if (preg_match('/\.pdf$/', $url)) {
            if (!in_array($url, $this->memory, true)) {
                $text = pdf_to_text($url);
                $doc_id = insert_document($url, $title);
                $lines = explode("\n", $text);
                foreach ($lines as $line) {
                    if ($line != NULL) {
                        $words = analyze($line);
                        foreach ($words as $word) {
                            insert_word($word, $doc_id);
                        }
                    }
                }
                // update last_index datetime
                index_finished($doc_id);
                echo 'done: ' . $depth . ' ' . $url . "\n";
            } else {
                echo 'pass: ' . $depth . ' ' . $url . "\n";
            }
            return;
        }

        $html = get_html($url);
        if ($html == NULL) {
            return;
        }

        if (!in_array($html['url'], $this->memory, true)) {
            $text = html_to_text($html);
            if ($html['title'] != "") {
                $title = $html['title'];
            }
            $doc_id = insert_document($html['url'], $title);
            array_push($this->memory, $html['url']);

            $lines = explode("\n", $text);
            foreach ($lines as $line) {
                if ($line != NULL) {
                    $words = analyze($line);
                    foreach ($words as $word) {
                        insert_word($word, $doc_id);
                    }
                }
            }
            // update last_index datetime
            index_finished($doc_id);
            echo 'done: ' . $depth . ' ' .  $html['url'] . "\n";
        } else {
            echo 'pass: ' . $depth . ' ' .  $html['url'] . "\n";
        }


        if ($depth < 1) {
            $absolute_paths = get_absolute_paths($html);
            foreach ($absolute_paths as $absolute_path) {
                $url = "https://www.chorkleines.com" . $absolute_path['path'];
                $this->crawler($url, $absolute_path['text'], $depth + 1);
            }
        }
        return;
    }
}

File: crawler/Crawler.py
from document_to_text import document_to_text
from file_to_text import file_to_text
from html_to_text import html_to_text, html_title
from find_url import find_url
from get_header import get_header
from get_html import get_html
import json
from collections import OrderedDict
import pprint
from mecab import mecab
import requests
import datetime
from mysql_connect import mysql_connect
import os


class Crawler():

    def __init__(self, url):
        self.memory = []
        self.crawler(url, "", 0)

    def __del__(self):
        self.conn.close()
        self.server.stop()

    def crawler(self, url, title, depth):
        target = datetime.datetime.now() - datetime.timedelta(days=30)
        target = target.strftime('%Y/%m/%d %H:%M:%S')
        query = "SELECT * FROM documents WHERE last_index IS NOT NULL AND last_index > %s"
        res = mysql_connect()
        self.server = res[0]
        self.conn = res[1]
        cur = self.conn.cursor()
        cur.execute(query, (target,))
        cur.close()
        for row in cur:
            self.memory.append(row[1])

        if "chorkleines.com/member/" not in url:
            return
        elif "chorkleines.com/member/bbs/" in url:
            return
        elif "chorkleines.com/member/download/18/pdf_search/" in url:
            return
        elif "chorkleines.com/member/download/18/scoredb/" in url:
            return
        elif "chorkleines.com/member/download/18/past_exam/" in url:
            return
        elif "chorkleines.com/member/wiki/" in url:
            return
        elif "chorkleines.com/member/kleines_search/" in url:
            return

        if url.endswith((".pdf", ".doc", ".docx")):
            if url not in self.memory:
                text = document_to_text(url)
                if text is None:
                    print("404: " + url)
                    return
                doc_id = self.insert_document(url, title)
                lines = text.splitlines()
                for line in lines:
                    if line != "":
                        line_words = mecab(line)
                        for line_word in line_words:
                            self.insert_word(line_word['text'], doc_id)
                line_words = mecab(title)
                for line_word in line_words:
                    self.insert_word(line_word['text'], doc_id)
                self.insert_done(doc_id)
                print("done: " + url)
            else:
                print("pass: " + url)
            return
        elif url.endswith((".csv", ".txt")):
            if url not in self.memory:
                text = file_to_text(url)
                if text is None:
                    print("404: " + url)
                    return
                doc_id = self.insert_document(url, title)
                lines = text.splitlines()
                for line in lines:
                    if line != "":
                        line_words = mecab(line)
                        for line_word in line_words:
                            self.insert_word(line_word['text'], doc_id)
                line_words = mecab(title)
                for line_word in line_words:
                    self.insert_word(line_word['text'], doc_id)
                self.insert_done(doc_id)
                print("done: " + url)
            else:
                print("pass: " + url)
            return
        elif url.endswith((".mp3", ".mp4", ".midi", ".mid", ".wav", ".zip", ".tar", ".gz", ".tgz", ".jpeg", ".jpg", ".png", ".xlsx", ".xls", ".pptx", ".ppt", ".mscz")):
            if url not in self.memory:
                if get_header(url) is None:
                    print("404: " + url)
                    return
                doc_id = self.insert_document(url, title)
                line_words = mecab(title)
                for line_word in line_words:
                    self.insert_word(line_word['text'], doc_id)
                self.insert_done(doc_id)
                print("done: " + url)
            else:
                print("pass: " + url)
            return
        elif url.endswith(("css", "js")):
            return

        data = get_html(url)
        if data is None:
            return
        url = data[0]
        html = data[1]

        if url not in self.memory:
            text = html_to_text(html)
            title_tmp = html_title(html)
            if title_tmp != "":
                title = title_tmp
            doc_id = self.insert_document(url, title)
            lines = text.splitlines()
            for line in lines:
                if line != "":
                    line_words = mecab(line)
                    for line_word in line_words:
                        self.insert_word(line_word['text'], doc_id)
            line_words = mecab(title)
            for line_word in line_words:
                self.insert_word(line_word['text'], doc_id)
            self.insert_done(doc_id)
            print("done: " + url)
        else:
            print("pass: " + url)
            # return

        if depth < 3:
            links = find_url(data)
            for link in links:
                self.crawler(link["href"], link["text"], depth + 1)

        return

    def insert_document(self, url, title):
        if url.endswith(
            ("mp3",
             "mp4",
             "midi",
             "mid",
             "wav",
             "zip",
             "tar",
             "gz",
             "tgz",
             "jpeg",
             "jpg",
             "png",
             "xlsx",
             "xls",
             "pptx",
             "ppt",
             " mscz",
             ".pdf",
             ".doc",
             ".docx",
             "csv",
             "txt")):
            filetype = os.path.splitext(url)[1][1:]
        else:
            filetype = ''

        query = """INSERT INTO documents (url, title, filetype)
            SELECT * FROM (SELECT %s AS url, %s AS title, %s AS filetype) AS tmp
            WHERE NOT EXISTS (
                SELECT url FROM documents WHERE url = %s
            ) LIMIT 1
            """
        cur = self.conn.cursor()
        url = self.conn.escape(url)
        title = self.conn.escape(title)
        filetype = self.conn.escape(filetype)
        cur.execute(query, (url, title, filetype, url))
        self.conn.commit()
        cur.close()

        query = "SELECT id FROM documents WHERE url = %s"
        cur = self.conn.cursor()
        url = self.conn.escape(url)
        cur.execute(query, (url,))
        cur.close()

        for row in cur:
            doc_id = row[0]
        doc_id = str(doc_id)
        doc_id = doc_id.zfill(10)

        query = "SHOW COLUMNS FROM inverted_index"
        cur = self.conn.cursor()
        cur.execute(query)
        cur.close()

        columns = []
        for row in cur:
            columns.append(row[0])

        if doc_id not in columns:
            query = "ALTER TABLE inverted_index ADD COLUMN `%s` int(5) DEFAULT '0'"
            cur = self.conn.cursor()
            doc_id = self.conn.escape(doc_id)
            cur.execute(query % (doc_id,))
            self.conn.commit()
            cur.close()

        query = "UPDATE inverted_index SET `%s` = 0"
        cur = self.conn.cursor()
        doc_id = self.conn.escape(doc_id)
        cur.execute(query % (doc_id,))
        self.conn.commit()
        cur.close()

        return doc_id

    def insert_word(self, text, doc_id):
        query = """INSERT INTO inverted_index(word)
            SELECT * FROM(SELECT %s) AS tmp
            WHERE NOT EXISTS(
                SELECT word FROM inverted_index WHERE word=%s
            ) LIMIT 1"""
        cur = self.conn.cursor()
        text = self.conn.escape(text)
        cur.execute(query, (text, text))
        self.conn.commit()
        cur.close()

        query = "UPDATE inverted_index SET `%s` = `%s` + 1 WHERE word =%s"
        cur = self.conn.cursor()
        doc_id = self.conn.escape(doc_id)
        text = self.conn.escape(text)
        cur.execute(query % (doc_id, doc_id, text))
        self.conn.commit()
        cur.close()

        return

    def insert_done(self, doc_id):
        query = "UPDATE documents SET last_index = NOW() WHERE id=%s"
        cur = self.conn.cursor()
        doc_id = self.conn.escape(doc_id)
        cur.execute(query, (doc_id,))
        self.conn.commit()
        cur.close()

        return


Crawler("https://www.chorkleines.com/member/download/18/")
